network/src/transport.rs

```diff
@@ -397,3 +420,92 @@ pub fn build_tcp_transport(
         .with_timeout(TRANSPORT_TIMEOUT)
         .boxed()
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        protocols::wire::handshake::v1::{HandshakeMsg, MessagingProtocolVersion},
+        transport::perform_handshake,
+        ProtocolId,
+    };
+    use futures::{executor::block_on, future::join};
+    use libra_config::network_id::NetworkId;
+    use libra_network_address::NetworkAddress;
+    use libra_types::PeerId;
+    use memsocket::MemorySocket;
+    use netcore::transport::ConnectionOrigin;
+
+    fn build_test_connection() -> (MemorySocket, MemorySocket) {
+        MemorySocket::new_pair()
+    }
+
+    #[test]
+    fn handshake_network_id_mismatch() {
+        let peer_id_1 = PeerId::from_hex_literal("0x00").unwrap();
+        let peer_id_2 = PeerId::from_hex_literal("0x01").unwrap();
+        let network_addr_1 = NetworkAddress::mock();
+        let network_addr_2 = NetworkAddress::mock();
+        let (outbound, inbound) = build_test_connection();
+        let validator_network = NetworkId::Validator;
+        let public_network = NetworkId::Public;
+
+        // Create client and server handshake messages.
+        let mut server_handshake = HandshakeMsg::new(validator_network);
+        server_handshake.add(
+            MessagingProtocolVersion::V1,
+            [
+                ProtocolId::ConsensusDirectSend,
+                ProtocolId::MempoolDirectSend,
+            ]
+            .iter()
+            .into(),
+        );
+        let mut client_handshake = HandshakeMsg::new(public_network);
+        client_handshake.add(
+            MessagingProtocolVersion::V1,
+            [ProtocolId::ConsensusRpc, ProtocolId::ConsensusDirectSend]
+                .iter()
+                .into(),
+        );
+
+        let server = async move {
+            perform_handshake(
+                peer_id_1,
+                inbound,
+                network_addr_1,
+                ConnectionOrigin::Inbound,
+                &server_handshake,
+            )
+            .await
+            .expect_err(
+                format!(
+                    "network_ids don't match own: {:?} received: {:?}",
+                    NetworkId::Validator,
+                    NetworkId::Public
+                )
+                .as_str(),
+            );
```

Comment:
probably could just do an unwrap_err here, unless you think we get additional signal from this? I don't think we do

Reply:
I think the idea was to ensure we aren't just getting some random other error